e2e-polybft/framework/test-cluster.go

```diff
@@ -554,3 +591,185 @@ func (c *TestCluster) InitSecrets(prefix string, count int) ([]types.Address, er
+	targetAddr := ethgo.Address(target)
+	return t.SendTxn(tt, sender, &ethgo.Transaction{To: &targetAddr, Value: value})
+}
+
+func (t *TestCluster) MethodTxn(tt *testing.T, sender ethgo.Key, target types.Address, input []byte) *TestTxn {
+	targetAddr := ethgo.Address(target)
+	return t.SendTxn(tt, sender, &ethgo.Transaction{To: &targetAddr, Input: input})
+}
+
+// SendTxn sends a transaction
+func (t *TestCluster) SendTxn(tt *testing.T, sender ethgo.Key, txn *ethgo.Transaction) *TestTxn {
+	// since we might use get nonce to query the latest nonce and that value is only
+	// updated if the transaction is on the pool, it is recommended to lock the whole
+	// execution in case we send multiple transactions from the same account and we expect
+	// to get a sequential nonce order.
+	t.sendTxnLock.Lock()
+	defer t.sendTxnLock.Unlock()
+
+	client, err := jsonrpc.NewClient(t.Servers[0].JSONRPCAddr())
+	require.NoError(tt, err)
+
+	// initialize transaction values if not set
+	if txn.Nonce == 0 {
+		nonce, err := client.Eth().GetNonce(sender.Address(), ethgo.Latest)
+		require.NoError(tt, err)
+
+		txn.Nonce = nonce
+	}
+	if txn.GasPrice == 0 {
+		txn.GasPrice = defaultGasPrice
+	}
+	if txn.Gas == 0 {
+		txn.Gas = defaultGasLimit
+	}
+
+	chainID, err := client.Eth().ChainID()
+	require.NoError(tt, err)
+
+	signer := wallet.NewEIP155Signer(chainID.Uint64())
+	signedTxn, err := signer.SignTx(txn, sender)
+	require.NoError(tt, err)
+
+	txnRaw, err := signedTxn.MarshalRLPTo(nil)
+	require.NoError(tt, err)
+
+	hash, err := client.Eth().SendRawTransaction(txnRaw)
+	require.NoError(tt, err)
+
+	tTxn := &TestTxn{
+		client: client.Eth(),
+		txn:    txn,
+		hash:   hash,
+	}
+	return tTxn
+}
+
+type TestTxn struct {
+	client  *jsonrpc.Eth
+	hash    ethgo.Hash
+	txn     *ethgo.Transaction
+	receipt *ethgo.Receipt
+}
+
+// Txn returns the raw transaction that was sent
+func (t *TestTxn) Txn() *ethgo.Transaction {
+	return t.txn
+}
+
+// Receipt returns the receipt of the transaction
+func (t *TestTxn) Receipt() *ethgo.Receipt {
+	return t.receipt
+}
+
+const (
+	statusFailed  = uint64(0)
+	statusSucceed = uint64(1)
+)
+
+// Succeed returns whether the transaction succeed and it was not reverted
+func (t *TestTxn) Succeed() bool {
+	return t.receipt.Status == statusSucceed
+}
+
+// Failed returns whether the transaction failed
+func (t *TestTxn) Failed() bool {
+	return t.receipt.Status == statusFailed
+}
+
+// Reverted returns whether the transaction failed and was reverted consuming
+// all the gas from the call
+func (t *TestTxn) Reverted() bool {
+	return t.receipt.Status == statusFailed && t.txn.Gas == t.receipt.GasUsed
+}
+
+// Wait waits for the transaction to be executed
+func (t *TestTxn) Wait() error {
+	return t.WaitWithDuration(1 * time.Minute)
+}
+
+// WaitWithDuration waits for the transaction to be executed with a given timeout
+func (t *TestTxn) WaitWithDuration(timeout time.Duration) error {
+	for {
```

Comment:
Why do you need for there?

Reply:
This was used before because I thought one of the transactions would never be executed. I will merge this with `Wait` and it can be introduced later if necessary.

Reply:
Just notice, that timeout will be recreated for every loop step(it's not a timeout for the whole function).